src/main/java/duke/task/TaskManager.java

```diff
@@ -0,0 +1,252 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.ArrayList;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import duke.DukeException;
+import duke.parser.DateHandler;
+
+/**
+ * Manager class to manage and keep track of all the actions being performed on the task.
+ */
+public class TaskManager {
+    private static final String listingResponse = "Here are the tasks in your list:";
+    //Strings for marking and unmarking
+    private static final String RESPONSE_MARK = "Nice! I've marked this task as done:";
+    private static final String RESPONSE_UMARK = "OK, I've marked this task as not done yet:";
+    private static final String RESPONSE_REMOVE = "Noted. I've removed this task";
+    //String and variables for task
+    private static final String RESPONSE_ADD = "Got it. I've added this task:";
+    private static final String RESPONSE_FIND = "Here are the matching tasks in your list";
+    private ArrayList<Task> items;
+    private boolean hasChanged = false;
+
+
+    /**
+     * Creates a manager.
+     */
+    public TaskManager() {
+        this.items = new ArrayList<>();
+    }
+
+    /**
+     * Adds task to the current manager list.
+     *
+     * @param item A task to add.
+     */
+    public void addItem(Task item) {
+        items.add(item);
+    }
+
+    /**
+     * Adds a Task based on the possible Actions and instruction to the current manage
+     *
+     * @param options     A valid Action to perform.
+     * @param instruction A string value of the commands.
+     * @return An ArrayList of string of the values to output to the Ui.
+     * @throws DukeException Invalid processing of the items.
+     */
+    public ArrayList<String> addTask(Actions options, String instruction) throws DukeException {
+        Task item;
+        String description;
+        String by;
+        String from;
+        Pattern deadlineFormat = Pattern.compile("(?<description>.+)\\s?((?i)/by)(?<by>.+)");
```

Comment:
wow, this is really impressive! it never crossed my mind to implement it, but after seeing your code, i'm definitely going to give it a try.